### PR TITLE
Configure: support KOKKOS when found with PETSc

### PR DESCRIPTION
**Pull request by stefanozampini**
This still require grabbing the compile flags and some special CUDA handling?

I'm not a CMAKE expert. I profoundly hate it. So if any CMAKE expert wants to pitch in and provide comments/suggestions/improvements, please go ahead @luca-heltai @bangerth

### CODE REVIEW

**Comment by luca-heltai** [comment]
I think @tamiko is our expert here. :)

**Comment by masterleinad** [comment]
Does `PETSc` bundle `Kokkos` or does it link to an external installation?

**Comment by stefanozampini** [comment]
We don't ship Kokkos source code. We link against an existing installation (that can also be automatically downloaded and built during PETSc configure)

File: cmake/modules/FindDEAL_II_KOKKOS.cmake [hunk 1]
```diff
@@ -26,7 +26,7 @@ set(KOKKOS_DIR "" CACHE PATH "An optional hint to a Kokkos installation")
 set_if_empty(KOKKOS_DIR "$ENV{KOKKOS_DIR}")
 
 
-if(DEAL_II_TRILINOS_WITH_KOKKOS)
+if(DEAL_II_TRILINOS_WITH_KOKKOS OR DEAL_II_PETSC_WITH_KOKKOS)
```
Comment (masterleinad) [inline cmake/modules/FindDEAL_II_KOKKOS.cmake:29]:
```suggestion
if(DEAL_II_TRILINOS_WITH_KOKKOS)
```

**Review by masterleinad — COMMENTED** [review]
> We don't ship Kokkos source code. We link against an existing installation (that can also be automatically downloaded and built during PETSc configure)

In that case, I would rather require that we don't allow using the bundled `Kokkos` package but require `Kokkos` to be found as an external package and not import `Kokkos` flags from `PETSc` but rather from `Kokkos` directly. We treat `Trilinos` differently because they can't yet use an external `Kokkos` installation.

This way, we could also allow building `deal.II` with `PETSc` with `Kokkos` and `Trilinos` with `Kokkos` assuming that `PETSc` could use `Trilinos`' bundled `Kokkos` installation.

File: cmake/modules/FindDEAL_II_KOKKOS.cmake (hunk 1 above)
Comment (stefanozampini) [reply cmake/modules/FindDEAL_II_KOKKOS.cmake:29]: I'm ok with this change. We probably need more automation and to provide hints to `find_package` to look into PETSc makefile variables `KOKKOS_INCLUDE` and `KOKKOS_LIB` https://github.com/dealii/dealii/blob/a35e4628710c7d165aaf6e0d4ab00b6f6f1a6ef1/cmake/modules/FindDEAL_II_PETSC.cmake#L136

**Comment by tamiko** [comment]
A couple of thoughts:
 - I think that this PR is a good starting point for a sanity check: force external kokkos if PETSc is configured with Kokkos. We already do something similar for Trilinos (Similarly to our MPI sanity checks we could make sure that we configure against one external kokkos).
 - For actually finding a Kokkos installation "bundled" with PETSc we should make sure that the `FindDEAL_II_KOKKOS` module has the right search paths (via appropriate hints). I am happy to look into this in the very near future:

That said: I really want to get the CMake refactoring done now and I suggest the we hold off with changes to external features for about 2-4 weeks until said refactoring is done. Otherwise we will have nasty inflight collisions. Or have to fix problems twice.

**Comment by tamiko** [comment]
Depends on #12374

**Comment by stefanozampini** [comment]
@tamiko This small change allows me to build and run using my PETSc install with KOKKOS dependency, otherwise there are a bunch of errors at runtime since PETSc and deal.II uses a different KOKKOS version. Can we merge this as is for now?

**Review by tamiko — APPROVED** [review]

**Comment by tamiko** [comment]
@stefanozampini Fair enough.

**Pull request merged**